File: tests/netDxf.Conformance/RawHandleIndexTests.cs
using netDxf;
using netDxf.Entities;
using netDxf.Header;
using netDxf.IO;

namespace NetDxf.Conformance;

internal static partial class Program
{
    private static IEnumerable<DxfVersion> HandleProfiles => new[] { DxfVersion.AutoCad12, DxfVersion.AutoCad13, DxfVersion.AutoCad14 }.Concat(SupportedVersions);
    private static string HandleProfileName(DxfVersion version) => version switch
    {
        DxfVersion.AutoCad12 => "AC1009", DxfVersion.AutoCad13 => "AC1012", DxfVersion.AutoCad14 => "AC1014", _ => HeaderVersion(version)
    };
    private static void RegisterRawHandleIndexTests()
    {
        foreach (DxfVersion version in HandleProfiles)
        foreach (bool binary in new[] { false, true })
        {
            DxfVersion v = version; bool b = binary;
            Run($"handles/contexts/{v}/{b}", () => HandleContexts(v, b));
            foreach (int variant in Enumerable.Range(0, 7))
            {
                int k = variant;
                Run($"handles/diagnostics/{v}/{b}/{k}", () => HandleDiagnostics(v, b, k));
            }
        }
        foreach (DxfVersion version in SupportedVersions)
        foreach (bool binary in new[] { false, true })
        {
            DxfVersion v = version; bool b = binary;
            Run($"handles/typed-output/{v}/{b}", () => HandleTypedOutput(v, b));
        }
        Run("handles/budgets-cancellation-and-stale-records", HandleIndexGuards);
        Run("handles/deep-owner-chain-no-recursion", HandleDeepChain);
        Run("handles/header-and-xrecord-xdata-context", HandleHeaderAndXRecord);
        Run("handles/read-only-concurrent-lookup", HandleConcurrentLookup);
    }

    private static List<DxfTag> HandleFixture(DxfVersion version)
    {
        return new List<DxfTag>
        {
            new(0,"SECTION"),new(2,"HEADER"),new(9,"$ACADVER"),new(1,HandleProfileName(version)),
            new(9,"$DWGCODEPAGE"),new(3,"ANSI_1252"),new(9,"$HANDSEED"),new(5,"0000AB"),new(0,"ENDSEC"),
            new(0,"SECTION"),new(2,"TABLES"),new(0,"TABLE"),new(2,"DIMSTYLE"),new(5,"11"),
            new(0,"DIMSTYLE"),new(105,"000d"),DxfTag.CreateDimensionStyleArrowName("00aB"),new(330,"11"),
            new(100,"AcDbSymbolTableRecord"),new(100,"AcDbDimStyleTableRecord"),new(2,"Style"),
            new(0,"ENDTAB"),new(0,"ENDSEC"),
            new(0,"SECTION"),new(2,"ENTITIES"),new(0,"LINE"),new(5,"00aB"),
            new(102,"{ACAD_REACTORS"),new(330,"E"),new(102,"}"),
            new(102,"{ACAD_XDICTIONARY"),new(360,"C"),new(102,"}"),
            new(102,"{VENDOR"),new(5,"ab"),new(330,"DEADBEEF"),
            new(102,"{ACAD_REACTORS"),new(330,"C"),new(102,"}"),new(102,"}"),
            new(330,"10"),new(100,"AcDbEntity"),new(8,"0"),new(100,"AcDbLine"),
            new(10,1.0),new(20,2.0),new(30,3.0),new(11,4.0),new(21,5.0),new(31,6.0),
            new(330,"F"),new(340,"E"),new(350,"C"),new(360,"E"),new(390,"C"),new(480,"E"),
            new(320,"AB"),new(1001,"HANDLES_TEST"),new(1002,"{"),new(1005,"E"),new(1002,"}"),
            new(0,"ENDSEC"),new(0,"SECTION"),new(2,"OBJECTS"),
            new(0,"DICTIONARY"),new(5,"10"),new(330,"0"),new(100,"AcDbDictionary"),new(3,"Children"),new(350,"C"),
            new(0,"DICTIONARY"),new(5,"C"),new(330,"AB"),new(100,"AcDbDictionary"),new(3,"Entry"),new(360,"E"),
            new(0,"XRECORD"),new(5,"E"),new(330,"C"),new(100,"AcDbXrecord"),new(280,(short)1),
            new(330,"DEADBEEF"),new(102,"application payload, not a control group"),new(1,"data"),
            new(0,"ENDSEC"),new(0,"SECTION"),new(2,"APP_SECTION"),new(0,"CUSTOM"),new(5,"AB"),
            new(0,"ENDSEC"),new(0,"EOF")
        };
    }
    private static DxfRawDocument HandleRoundTrip(IEnumerable<DxfTag> tags, bool binary)
    {
        var authored = DxfRawDocument.Create(tags, binary);
        using var stream = new MemoryStream(); authored.Save(stream); stream.Position = 0;
        return DxfRawDocument.Load(stream);
    }
    private static void HandleContexts(DxfVersion version, bool binary)
    {
        var doc = HandleRoundTrip(HandleFixture(version), binary);
        using var before = new MemoryStream(); doc.Save(before);
        var index = DxfRawHandleIndex.Create(doc);
        Equal(1,index.FindDefinitions("ab").Count,"Header seed/vendor/unknown identity collision");
        Equal("LINE",index.FindDefinitions("000AB").Single().Record.Name,"Numeric lookup alias");
        Equal((string)doc.Tags[index.FindDefinitions("AB").Single().TagIndex].Value,index.FindDefinitions("AB").Single().Handle,"Decoded spelling lost");
        var authoredIndex=DxfRawHandleIndex.Create(DxfRawDocument.Create(HandleFixture(version)));
        Equal("00aB",authoredIndex.FindDefinitions("AB").Single().Handle,"Authored spelling lost");
        Equal((short)105,index.FindDefinitions("D").Single().Code,"DIMSTYLE group 105 identity");
        Check(index.Occurrences.All(x=> x.Code!=5 || x.Record?.Name!="DIMSTYLE"),"DIMBLK name was indexed as handle");
        var line = index.FindDefinitions("AB").Single().Record;
        var items = index.GetOccurrences(line);
        Equal(1,items.Count(x=>x.Role==DxfRawHandleRole.Owner),"Reactor/application pointer mistaken for owner");
        Equal("10",items.Single(x=>x.Role==DxfRawHandleRole.Owner).CanonicalHandle,"Wrong common owner");
        Equal("E",items.Single(x=>x.Role==DxfRawHandleRole.Reactor).CanonicalHandle,"Reactor lost");
        Equal("C",items.Single(x=>x.Role==DxfRawHandleRole.ExtensionDictionary).CanonicalHandle,"Extension dictionary lost");
        Equal(3,items.Count(x=>x.Role==DxfRawHandleRole.Opaque),"Nested/custom handles incorrectly interpreted");
        Check(items.Any(x=>x.Code==330 && x.Role==DxfRawHandleRole.SoftPointer && x.CanonicalHandle=="F"),"Subclass pointer mistaken for owner");
        Equal("HANDLES_TEST",items.Single(x=>x.Role==DxfRawHandleRole.XData).Context,"XData APPID context");
        Check(index.FindReferences("AB").All(x=>x.Role!=DxfRawHandleRole.Arbitrary),"Nontranslated arbitrary slot included in references");
        Equal(1,index.Occurrences.Count(x=>x.Role==DxfRawHandleRole.HeaderSeed),"Header seed misclassified");
        Equal(1,index.Diagnostics.Count(x=>x.Kind==DxfRawHandleDiagnosticKind.UnresolvedReference),"Opaque payload manufactured dangling pointers");
        Equal("F",index.Diagnostics.Single().Handle,"Unexpected structural diagnostic");
        foreach (var occurrence in index.Occurrences)
            Equal(occurrence.Handle,(string)doc.Tags[occurrence.TagIndex].Value,"Absolute occurrence location");
        using var after = new MemoryStream();doc.Save(after);
        Check(before.ToArray().SequenceEqual(after.ToArray()),"Indexing changed exact original bytes");
    }
    private static void HandleDiagnostics(DxfVersion version, bool binary, int variant)
    {
        var tags=HandleFixture(version);int at=tags.FindIndex(t=>t.Code==0 && Equals(t.Value,"LINE"));
        DxfRawHandleDiagnosticKind expected=variant switch
        {
            0=>DxfRawHandleDiagnosticKind.DuplicateIdentity,1=>DxfRawHandleDiagnosticKind.MultipleIdentities,
            2=>DxfRawHandleDiagnosticKind.NullIdentity,3=>DxfRawHandleDiagnosticKind.MultipleOwners,
            4=>DxfRawHandleDiagnosticKind.InvalidControlGroup,5=>DxfRawHandleDiagnosticKind.OwnerCycle,
            _=>DxfRawHandleDiagnosticKind.AmbiguousReference
        };
        if(variant==0||variant==6)tags.InsertRange(at,new DxfTag[]{new(0,"POINT"),new(5,variant==0?"AB":"E")});
        if(variant==1)tags.Insert(at+2,new(5,"222"));
        if(variant==2)tags[at+1]=new(5,"000");
        if(variant==3)tags.Insert(at+2,new(330,"E"));
        if(variant==4)tags.Insert(at+2,new(102,"}"));
        if(variant==5)
        {
            int owner=tags.FindIndex(at,t=>t.Code==330 && Equals(t.Value,"10"));tags[owner]=new(330,"C");
        }
        var index=DxfRawHandleIndex.Create(HandleRoundTrip(tags,binary));
        Check(index.Diagnostics.Any(x=>x.Kind==expected),"Expected handle diagnostic missing: "+expected);
        if(variant==0)Equal(2,index.FindDefinitions("00ab").Count,"Duplicate definition was overwritten");
        if(variant==5)Equal(2,index.Diagnostics.Count(x=>x.Kind==expected),"Owner cycle membership");
    }
    private static void HandleTypedOutput(DxfVersion version,bool binary)
    {
        var typed=new DxfDocument(version);typed.Entities.Add(new Line(Vector3.Zero,new Vector3(1,2,3)));
        using var stream=new MemoryStream();Check(typed.Save(stream,binary),"Typed fixture save");stream.Position=0;
        var doc=DxfRawDocument.Load(stream);var index=DxfRawHandleIndex.Create(doc);
        var line=doc.Sections.SelectMany(s=>s.Records).Single(r=>r.Name=="LINE");
        var own=index.GetOccurrences(line).Single(x=>x.Role==DxfRawHandleRole.Owner);
        Equal("BLOCK_RECORD",index.FindDefinitions(own.Handle).Single().Record.Name,"Typed common owner resolution");
        Check(!index.Diagnostics.Any(x=>x.Kind is DxfRawHandleDiagnosticKind.DuplicateIdentity or DxfRawHandleDiagnosticKind.InvalidControlGroup or DxfRawHandleDiagnosticKind.OwnerCycle),"Valid typed structure produced contradictory identity/control/owner graph");
    }
    private static void HandleIndexGuards()
    {
        var doc=DxfRawDocument.Create(HandleFixture(DxfVersion.AutoCad2018));var index=DxfRawHandleIndex.Create(doc);
        Throws<ArgumentNullException>(()=>DxfRawHandleIndex.Create(null!));
        Throws<ArgumentOutOfRangeException>(()=>new DxfRawHandleIndexOptions(0));
        Throws<ArgumentOutOfRangeException>(()=>new DxfRawHandleIndexOptions(1,0));
        Throws<InvalidDataException>(()=>DxfRawHandleIndex.Create(doc,new DxfRawHandleIndexOptions(1)));
        var bad=HandleFixture(DxfVersion.AutoCad2018);int at=bad.FindIndex(t=>t.Code==0&&Equals(t.Value,"LINE"));
        bad.InsertRange(at,new DxfTag[]{new(0,"POINT"),new(5,"AB")});
        Throws<InvalidDataException>(()=>DxfRawHandleIndex.Create(DxfRawDocument.Create(bad),new DxfRawHandleIndexOptions(1000,1)));
        using var cancel=new CancellationTokenSource();cancel.Cancel();
        Throws<OperationCanceledException>(()=>DxfRawHandleIndex.Create(doc,cancellationToken:cancel.Token));
        foreach(string handle in new[]{"", " 1", "1 ", "+1", "0x1", "G", "10000000000000000"})
            Throws<ArgumentException>(()=>index.FindDefinitions(handle));
        Throws<ArgumentNullException>(()=>index.FindReferences(null!));
        var other=doc.WithTags(doc.Tags);var stale=other.Sections.SelectMany(s=>s.Records).First();
        Throws<ArgumentException>(()=>index.GetOccurrences(stale));
        Throws<NotSupportedException>(()=>((IList<DxfRawHandleOccurrence>)index.Occurrences).Clear());
        Throws<NotSupportedException>(()=>((IList<DxfRawHandleDiagnostic>)index.Diagnostics).Clear());
        Throws<NotSupportedException>(()=>((IList<DxfRawHandleOccurrence>)index.FindDefinitions("AB")).Clear());
    }
    private static void HandleDeepChain()
    {
        var tags=new List<DxfTag>{new(0,"SECTION"),new(2,"HEADER"),new(9,"$ACADVER"),new(1,"AC1032"),new(0,"ENDSEC"),new(0,"SECTION"),new(2,"OBJECTS")};
        for(int i=12000;i>=1;i--)tags.AddRange(new DxfTag[]{new(0,"DICTIONARY"),new(5,i.ToString("X")),new(330,(i-1).ToString("X")),new(100,"AcDbDictionary")});
        tags.AddRange(new DxfTag[]{new(0,"ENDSEC"),new(0,"EOF")});
        var index=DxfRawHandleIndex.Create(DxfRawDocument.Create(tags));
        Equal(24000,index.Occurrences.Count,"Deep graph lost occurrences");Equal(0,index.Diagnostics.Count,"Deep acyclic graph diagnostics");
    }
    private static void HandleConcurrentLookup()
    {
        var doc=DxfRawDocument.Create(HandleFixture(DxfVersion.AutoCad2018));var index=DxfRawHandleIndex.Create(doc);
        Parallel.For(0,256,i=>{
            Equal("LINE",index.FindDefinitions("aB").Single().Record.Name,"Concurrent identity");
            Check(index.FindReferences("E").Count>1,"Concurrent pointers");
            var independent=DxfRawHandleIndex.Create(doc);
            Check(independent.Occurrences.Select(x=>x.TagIndex).SequenceEqual(index.Occurrences.Select(x=>x.TagIndex)),"Nondeterministic scan");
        });
    }
    private static void HandleHeaderAndXRecord()
    {
        var tags=HandleFixture(DxfVersion.AutoCad2018);
        int header=tags.FindIndex(t=>t.Code==0&&Equals(t.Value,"ENDSEC"));
        tags.InsertRange(header,new DxfTag[]{new(9,"$CUSTOM"),new(5,"AB"),new(9,"$CUSTOMARB"),new(320,"AB"),new(9,"$CMATERIAL"),new(347,"C")});
        int data=tags.FindIndex(t=>t.Code==102&&Equals(t.Value,"application payload, not a control group"));
        tags.InsertRange(data+2,new DxfTag[]{new(1001,"XREC_APP"),new(1005,"C")});
        var index=DxfRawHandleIndex.Create(DxfRawDocument.Create(tags));
        Equal(DxfRawHandleRole.Opaque,index.Occurrences.Single(x=>x.Record?.Name=="$CUSTOM").Role,"Unknown header identity-looking field inferred as reference");
        Equal(DxfRawHandleRole.Arbitrary,index.Occurrences.Single(x=>x.Record?.Name=="$CUSTOMARB").Role,"Header arbitrary handle inferred as translated reference");
        Equal(DxfRawHandleRole.HeaderReference,index.Occurrences.Single(x=>x.Record?.Name=="$CMATERIAL").Role,"Header material pointer lost");
        Equal(DxfRawHandleRole.XData,index.Occurrences.Single(x=>x.Context=="XREC_APP").Role,"XRECORD extension data confused with ordinary payload");
    }

}

File: javascript/tests/netDxf.Conformance/RawHandleIndexTests.js
// Port of tests/netDxf.Conformance/RawHandleIndexTests.cs; typed-output controls remain unported.
import { DxfTag, DxfRawDocument, DxfVersion, DxfRawHandleIndex, DxfRawHandleRole as R,
  DxfRawHandleDiagnosticKind as D, DxfRawHandleIndexOptions } from '../../index.js';
import * as E from '../../runtime/Errors.js';
import { Run, Check, Equal, Throws, SupportedVersions, VersionName, HeaderVersion, BooleanName } from './TestHarness.js';
import { SaveRaw, LoadRaw } from './RawDocumentTests.js';
export const HandleProfiles = [DxfVersion.AutoCad12, DxfVersion.AutoCad13, DxfVersion.AutoCad14, ...SupportedVersions];
export const HandleProfileName = HeaderVersion;
const T = (code, value) => new DxfTag(code, value);
export function RegisterRawHandleIndexTests() {
  for (const v of HandleProfiles) for (const b of [false, true]) {
    Run(`handles/contexts/${VersionName(v)}/${BooleanName(b)}`, () => HandleContexts(v, b));
    for (let k = 0; k < 7; k++)
      Run(`handles/diagnostics/${VersionName(v)}/${BooleanName(b)}/${k}`, () => HandleDiagnostics(v, b, k));
  }
  Run('handles/budgets-cancellation-and-stale-records', HandleIndexGuards);
  Run('handles/deep-owner-chain-no-recursion', HandleDeepChain);
  Run('handles/header-and-xrecord-xdata-context', HandleHeaderAndXRecord);
  Run('handles/read-only-concurrent-lookup', HandleConcurrentLookup);
}
export function HandleFixture(version) {
  return [T(0,'SECTION'),T(2,'HEADER'),T(9,'$ACADVER'),T(1,HandleProfileName(version)),
    T(9,'$DWGCODEPAGE'),T(3,'ANSI_1252'),T(9,'$HANDSEED'),T(5,'0000AB'),T(0,'ENDSEC'),
    T(0,'SECTION'),T(2,'TABLES'),T(0,'TABLE'),T(2,'DIMSTYLE'),T(5,'11'),
    T(0,'DIMSTYLE'),T(105,'000d'),DxfTag.CreateDimensionStyleArrowName('00aB'),T(330,'11'),
    T(100,'AcDbSymbolTableRecord'),T(100,'AcDbDimStyleTableRecord'),T(2,'Style'),T(0,'ENDTAB'),T(0,'ENDSEC'),
    T(0,'SECTION'),T(2,'ENTITIES'),T(0,'LINE'),T(5,'00aB'),
    T(102,'{ACAD_REACTORS'),T(330,'E'),T(102,'}'),T(102,'{ACAD_XDICTIONARY'),T(360,'C'),T(102,'}'),
    T(102,'{VENDOR'),T(5,'ab'),T(330,'DEADBEEF'),T(102,'{ACAD_REACTORS'),T(330,'C'),T(102,'}'),T(102,'}'),
    T(330,'10'),T(100,'AcDbEntity'),T(8,'0'),T(100,'AcDbLine'),
    T(10,1),T(20,2),T(30,3),T(11,4),T(21,5),T(31,6),
    T(330,'F'),T(340,'E'),T(350,'C'),T(360,'E'),T(390,'C'),T(480,'E'),T(320,'AB'),
    T(1001,'HANDLES_TEST'),T(1002,'{'),T(1005,'E'),T(1002,'}'),
    T(0,'ENDSEC'),T(0,'SECTION'),T(2,'OBJECTS'),
    T(0,'DICTIONARY'),T(5,'10'),T(330,'0'),T(100,'AcDbDictionary'),T(3,'Children'),T(350,'C'),
    T(0,'DICTIONARY'),T(5,'C'),T(330,'AB'),T(100,'AcDbDictionary'),T(3,'Entry'),T(360,'E'),
    T(0,'XRECORD'),T(5,'E'),T(330,'C'),T(100,'AcDbXrecord'),T(280,1),
    T(330,'DEADBEEF'),T(102,'application payload, not a control group'),T(1,'data'),
    T(0,'ENDSEC'),T(0,'SECTION'),T(2,'APP_SECTION'),T(0,'CUSTOM'),T(5,'AB'),T(0,'ENDSEC'),T(0,'EOF')];
}
export function HandleRoundTrip(tags, binary) { return LoadRaw(SaveRaw(DxfRawDocument.Create(tags, binary))); }
export function HandleContexts(v, b) {
  const doc = HandleRoundTrip(HandleFixture(v), b), before = SaveRaw(doc), index = DxfRawHandleIndex.Create(doc);
  Equal(1, index.FindDefinitions('ab').Count);
  Equal('LINE', index.FindDefinitions('000AB')[0].Record.Name);
  const definition = index.FindDefinitions('AB')[0];
  Equal(doc.Tags[definition.TagIndex].Value, definition.Handle);
  Equal('00aB', DxfRawHandleIndex.Create(DxfRawDocument.Create(HandleFixture(v))).FindDefinitions('AB')[0].Handle);
  Equal(105, index.FindDefinitions('D')[0].Code);
  Check(index.Occurrences.every(x => x.Code !== 5 || x.Record?.Name !== 'DIMSTYLE'));
  const items = index.GetOccurrences(definition.Record);
  Equal(1, items.filter(x => x.Role === R.Owner).length);
  Equal('10', items.find(x => x.Role === R.Owner).CanonicalHandle);
  Equal('E', items.find(x => x.Role === R.Reactor).CanonicalHandle);
  Equal('C', items.find(x => x.Role === R.ExtensionDictionary).CanonicalHandle);
  Equal(3, items.filter(x => x.Role === R.Opaque).length);
  Check(items.some(x => x.Code === 330 && x.Role === R.SoftPointer && x.CanonicalHandle === 'F'));
  Equal('HANDLES_TEST', items.find(x => x.Role === R.XData).Context);
  Check(index.FindReferences('AB').every(x => x.Role !== R.Arbitrary));
  Equal(1, index.Occurrences.filter(x => x.Role === R.HeaderSeed).length);
  Equal(1, index.Diagnostics.filter(x => x.Kind === D.UnresolvedReference).length);
  Equal('F', index.Diagnostics[0].Handle); Equal(1, index.Diagnostics.Count);
  for (const item of index.Occurrences) Equal(item.Handle, doc.Tags[item.TagIndex].Value);
  Equal(before, SaveRaw(doc));
}
export function HandleDiagnostics(v, b, variant) {
  const tags = HandleFixture(v), at = tags.findIndex(t => t.Code === 0 && t.Value === 'LINE');
  const expected = [D.DuplicateIdentity, D.MultipleIdentities, D.NullIdentity, D.MultipleOwners,
    D.InvalidControlGroup, D.OwnerCycle, D.AmbiguousReference][variant];
  if (variant === 0 || variant === 6) tags.splice(at, 0, T(0,'POINT'), T(5, variant === 0 ? 'AB' : 'E'));
  if (variant === 1) tags.splice(at + 2, 0, T(5,'222'));
  if (variant === 2) tags[at + 1] = T(5,'000');
  if (variant === 3) tags.splice(at + 2, 0, T(330,'E'));
  if (variant === 4) tags.splice(at + 2, 0, T(102,'}'));
  if (variant === 5) tags[tags.findIndex((t, i) => i >= at && t.Code === 330 && t.Value === '10')] = T(330,'C');
  const index = DxfRawHandleIndex.Create(HandleRoundTrip(tags, b));
  Check(index.Diagnostics.some(x => x.Kind === expected), 'Missing diagnostic ' + expected);
  if (variant === 0) Equal(2, index.FindDefinitions('00ab').Count);
  if (variant === 5) Equal(2, index.Diagnostics.filter(x => x.Kind === expected).length);
}
export function HandleIndexGuards() {
  const doc = DxfRawDocument.Create(HandleFixture(DxfVersion.AutoCad2018)), index = DxfRawHandleIndex.Create(doc);
  Throws(E.ArgumentNullException, () => DxfRawHandleIndex.Create(null));
  Throws(E.ArgumentOutOfRangeException, () => new DxfRawHandleIndexOptions(0));
  Throws(E.ArgumentOutOfRangeException, () => new DxfRawHandleIndexOptions(1, 0));
  Throws(E.InvalidDataException, () => DxfRawHandleIndex.Create(doc, new DxfRawHandleIndexOptions(1)));
  const bad = HandleFixture(DxfVersion.AutoCad2018), at = bad.findIndex(t => t.Code === 0 && t.Value === 'LINE');
  bad.splice(at, 0, T(0,'POINT'), T(5,'AB'));
  Throws(E.InvalidDataException, () => DxfRawHandleIndex.Create(DxfRawDocument.Create(bad), new DxfRawHandleIndexOptions(1000, 1)));
  Throws(E.OperationCanceledException, () => DxfRawHandleIndex.Create(doc, null, { aborted: true }));
  for (const handle of ['', ' 1', '1 ', '+1', '0x1', 'G', '10000000000000000'])
    Throws(E.ArgumentException, () => index.FindDefinitions(handle));
  Throws(E.ArgumentNullException, () => index.FindReferences(null));
  const stale = doc.WithTags(doc.Tags).Sections.flatMap(s => s.Records)[0];
  Throws(E.ArgumentException, () => index.GetOccurrences(stale));
  for (const items of [index.Occurrences, index.Diagnostics, index.FindDefinitions('AB')])
    Throws(TypeError, () => items.splice(0));
}
export function HandleDeepChain() {
  const tags = [T(0,'SECTION'),T(2,'HEADER'),T(9,'$ACADVER'),T(1,'AC1032'),T(0,'ENDSEC'),T(0,'SECTION'),T(2,'OBJECTS')];
  for (let i = 12000; i >= 1; i--)
    tags.push(T(0,'DICTIONARY'), T(5,i.toString(16).toUpperCase()), T(330,(i-1).toString(16).toUpperCase()), T(100,'AcDbDictionary'));
  tags.push(T(0,'ENDSEC'), T(0,'EOF'));
  const index = DxfRawHandleIndex.Create(DxfRawDocument.Create(tags));
  Equal(24000, index.Occurrences.Count); Equal(0, index.Diagnostics.Count);
}
export async function HandleConcurrentLookup() {
  const doc = DxfRawDocument.Create(HandleFixture(DxfVersion.AutoCad2018)), index = DxfRawHandleIndex.Create(doc);
  // Native JS objects are isolate-local. This checks interleaved readers, not shared CLR threads.
  await Promise.all(Array.from({ length: 256 }, async () => {
    await Promise.resolve();
    Equal('LINE', index.FindDefinitions('aB')[0].Record.Name); Check(index.FindReferences('E').Count > 1);
    const independent = DxfRawHandleIndex.Create(doc);
    Equal(index.Occurrences.map(x => x.TagIndex), independent.Occurrences.map(x => x.TagIndex));
  }));
}
export function HandleHeaderAndXRecord() {
  const tags = HandleFixture(DxfVersion.AutoCad2018), header = tags.findIndex(t => t.Code === 0 && t.Value === 'ENDSEC');
  tags.splice(header, 0, T(9,'$CUSTOM'),T(5,'AB'),T(9,'$CUSTOMARB'),T(320,'AB'),T(9,'$CMATERIAL'),T(347,'C'));
  const data = tags.findIndex(t => t.Code === 102 && t.Value === 'application payload, not a control group');
  tags.splice(data + 2, 0, T(1001,'XREC_APP'),T(1005,'C'));
  const index = DxfRawHandleIndex.Create(DxfRawDocument.Create(tags));
  Equal(R.Opaque, index.Occurrences.find(x => x.Record?.Name === '$CUSTOM').Role);
  Equal(R.Arbitrary, index.Occurrences.find(x => x.Record?.Name === '$CUSTOMARB').Role);
  Equal(R.HeaderReference, index.Occurrences.find(x => x.Record?.Name === '$CMATERIAL').Role);
  Equal(R.XData, index.Occurrences.find(x => x.Context === 'XREC_APP').Role);
}
